src/controller/python/ChipDeviceController-ScriptDevicePairingDelegate.cpp

```diff
@@ -66,6 +66,28 @@ void ScriptDevicePairingDelegate::SetCommissioningStatusUpdateCallback(
     mOnCommissioningStatusUpdateCallback = callback;
 }
 
+void ScriptDevicePairingDelegate::OnStatusUpdate(DevicePairingDelegate::Status status)
+{
+    switch (status)
+    {
+    case DevicePairingDelegate::Status::SecurePairingSuccess:
+        ChipLogProgress(Zcl, "Secure Pairing Success");
+        break;
+    case DevicePairingDelegate::Status::SecurePairingFailed:
+        ChipLogError(Zcl, "Secure Pairing Failed");
+        if (mOnPairingCompleteCallback != nullptr)
+        {
+            // Incorrect state is the same error that chip-tool sends. We are also
+            // leveraging the on pairing complete callback to indicate that pairing
+            // has failed.
+            mOnPairingCompleteCallback(ToPyChipError(CHIP_ERROR_INCORRECT_STATE));
```

Comment:
If PASE actually does fail, do we not get both `OnPairingComplete` with error and `OnStatusUpdate` with error?

And if we do, do we need to null out `mOnPairingCompleteCallback` here, or in some way flag that we called the callback for this pairing attempt or something?  Or is the double-call into the callback in that case OK?

Reply:
>If PASE actually does fail, do we not get both OnPairingComplete with error and OnStatusUpdate with error?

Yes we would. The alternative solution (that I actually originally had) was calling the `OnPairingComplete` inside `SetUpCodePairer::OnDeviceDiscoveredTimeoutCallback`.

Another idea that just came to mind now is that inside `DeviceCommissioner::RendezvousCleanup` we could still call `mPairingDelegate->OnPairingComplete` if the error code is `CHIP_ERROR_TIMEOUT`.

Is either of these two solution more preferable for you?

>And if we do, do we need to null out mOnPairingCompleteCallback here, or in some way flag that we called the callback for this pairing attempt or something? Or is the double-call into the callback in that case OK?

Nulling `mOnPairingCompleteCallback` is not going to work at this callback is only setup once during init.

Good catch, There is a problem if this is called twice. Depending on you answer to my above response, possibly changes to this file are not going to be needed and we might not need a guard.

Reply:
I think any solution that changes what notifications are delivered will be a breaking change for existing SDK consumers and likely not something we want to do.  I'd like us to deliver sane notifications from CHIPDeviceController, but those would need to be alongside the existing broken ones to avoid breaking everyone.

So whatever solution we adopt here has to be contained within `src/controller/python/ChipDeviceController-ScriptDevicePairingDelegate.cpp` and work with the existing notifications as they are.

Reply:
Looks good, thanks!